swarm/storage/chunker.go

```diff
@@ -245,107 +245,79 @@ func (self *TreeChunker) split(depth int, treeSize int64, key Key, data io.Reade
-				self.hashChunk(hasher, job, chunkC, storeWg)
+
+				h, err := putter.Put(job.chunk)
+				if err != nil {
+					errC <- err
+					return
+				}
+				copy(job.key, h)
+				job.parentWg.Done()
 			case <-quitC:
 				return
 			}
 		}
 	}()
 }
 
-// The treeChunkers own Hash hashes together
-// - the size (of the subtree encoded in the Chunk)
-// - the Chunk, ie. the contents read from the input reader
-func (self *TreeChunker) hashChunk(hasher SwarmHash, job *hashJob, chunkC chan *Chunk, storeWg *sync.WaitGroup) {
-	hasher.ResetWithLength(job.chunk[:8]) // 8 bytes of length
-	hasher.Write(job.chunk[8:])           // minus 8 []byte length
-	h := hasher.Sum(nil)
-
-	newChunk := NewChunk(h, nil)
-	newChunk.SData = job.chunk
-	newChunk.Size = job.size
-
-	// report hash of this chunk one level up (keys corresponds to the proper subslice of the parent chunk)
-	copy(job.key, h)
-	// send off new chunk to storage
-	job.parentWg.Done()
-
-	if chunkC != nil {
-		//NOTE: this increases the chunk count even if the local node already has this chunk;
-		//on file upload the node will increase this counter even if the same file has already been uploaded
-		//So it should be evaluated whether it is worth keeping this counter
-		//and/or actually better track when the chunk is Put to the local database
-		//(which may question the need for disambiguation when a completely new chunk has been created
-		//and/or a chunk is being put to the local DB; for chunk tracking it may be worth distinguishing
-		newChunkCounter.Inc(1)
-		chunkC <- newChunk
-		storeWg.Add(1)
-		go func() {
-			defer storeWg.Done()
-			<-newChunk.dbStored
-		}()
-	}
-}
-
-func (self *TreeChunker) Append(key Key, data io.Reader, chunkC chan *Chunk) (Key, func(), error) {
+func (self *TreeChunker) Append(key Key, data io.Reader, putGetter PutGetter) (Key, func(), error) {
 	return nil, nil, errAppendOppNotSuported
 }
 
 // LazyChunkReader implements LazySectionReader
 type LazyChunkReader struct {
-	key       Key         // root key
-	chunkC    chan *Chunk // chunk channel to send retrieve requests on
-	chunk     *Chunk      // size of the entire subtree
-	off       int64       // offset
-	chunkSize int64       // inherit from chunker
-	branches  int64       // inherit from chunker
-	hashSize  int64       // inherit from chunker
+	key Key // root key
+	// chunkC    chan *Chunk // chunk channel to send retrieve requests on
+	chunkData ChunkData // size of the entire subtree
+	off       int64     // offset
+	chunkSize int64     // inherit from chunker
+	branches  int64     // inherit from chunker
+	hashSize  int64     // inherit from chunker
 	depth     int
+	getter    Getter
 }
 
 // implements the Joiner interface
-func (self *TreeChunker) Join(key Key, chunkC chan *Chunk, depth int) LazySectionReader {
+func (self *TreeChunker) Join(key Key, getter Getter, depth int) LazySectionReader {
 	return &LazyChunkReader{
-		key:       key,
-		chunkC:    chunkC,
+		key: key,
+		// chunkC:    chunkC,
```

Comment:
remove commented line

Reply:
Fixed in f32d4ccca55c57de559174ba876b5bcc419a087e